wit-0.3.0-draft/types.wit

```
@@ -348,22 +306,28 @@ interface types {
-    ///
-    /// This headers resource is a child: it must be dropped before the parent
-    /// `request` is dropped, or its ownership is transferred to another
-    /// component by e.g. `handler.handle`.
     headers: func() -> headers;
 
-    /// Get the body associated with the Request, if any.
+    /// Get body of the Request.
     ///
-    /// This body resource is a child: it must be dropped before the parent
-    /// `request` is dropped, or its ownership is transferred to another
-    /// component by e.g. `handler.handle`.
-    body: func() -> option<body>;
-
-    /// Takes ownership of the `request` and returns the `headers`, `body`
-    /// and `request-options`, if any.
-    into-parts: static func(this: request) -> tuple<headers, option<body>, option<request-options>>;
+    /// Stream returned by this method represents the contents of the body.
+    /// Once the stream is reported as closed, callers should await the returned future
+    /// to determine whether the body was received successfully.
+    /// The future will only resolve after the stream is reported as closed.
+    ///
+    /// The stream and future returned by this method are children:
+    /// they should be closed or consumed before the parent `response`
+    /// is dropped, or its ownership is transferred to another component
+    /// by e.g. `handler.handle`.
+    ///
+    /// This method may be called multiple times.
+    ///
+    /// This method will return `none` if it is called while either:
+    /// - a stream or future returned by a previous call to this method is still open
+    /// - a stream returned by a previous call to this method has reported itself as closed
+    /// Thus there will always be at most one readable stream open for a given body.
+    /// Each subsequent stream picks up where the last stream left off, up until it is finished.
+    body: func() -> option<tuple<stream<u8>, future<result<option<trailers>, error-code>>>>;
```

I'm a bit unsure whether using `option` for this makes sense: it seems overloaded for request that just don't have a body, and lead to confusing silent failures.

Given that, ISTM the better, if slightly more onerous, signature would be
```wit
body: func() -> option<result<tuple<stream<u8>, future<result<option<trailers>, error-code>>>, body-in-use-error>>;
```

(Where `body-in-use-error` would be a new thing, where I don't have strong feelings about what it should look like, exactly.)

---

I could be missing it, but I couldn't find any prohibition against trailers in `GET` requests, so if there were an `option` wrapping the `stream<u8>`, I think it'd need to be inside the `tuple`.  However, I haven't been able to find a semantic distinction between `none` and "zero-length contents" inside RFC 9110, so I worry that by introducing this distinction, we might be creating something that content may incorrectly depend on.  (E.g., let's say a component decides to fail if the `option<stream<u8>>` is non-`none`, but now it's rejecting `some` zero-length stream which maybe it gets passed (in a multi-component composition).).

Considering the perf angle (which could be another reason for the `option`):
* As an optimization and convenience, I do think it'd make sense to have the `contents` parameter to `{request,response}.new` be `option<stream<u8>>` as Till suggested, but `none` could be spec-defined to be normalized to an empty stream internally.
* I expect guest code simply won't ask for the content stream for `GET` requests, so source-language stream objects won't be wastefully created.  The CABI overhead of creating closed-on-construction streams should also be light, I think.

So perhaps:
```wit
body: func() -> result<tuple<stream<u8>, future<result<option<trailers>, error-code>>>>;
```
?

---

Good catch: I hadn't considered that my proposal would prevent access to trailers as well.

With that, I agree with the proposed interface, with one slight exception: I do think this is a categorically different error from the ones represented by `error-code`, and that we should make it its own type. Is this perhaps even a pattern that'd make sense to lift up higher and use the same error type for in other places as well?

---

As currently proposed, the error type in this case is actually `unit`, e.g. in Rust it's `()`.
Here's the host implementation with the full Rust type: https://github.com/bytecodealliance/wasip3-prototyping/blob/ecd781215e650dc5eb4e6ec4b2fd60858c084654/crates/wasi-http/src/p3/host/types.rs#L669-L700

Does this address the concern?

---

Kinda? It's better than using `error-code` (sorry for missing that it's not), but it does seem very low-effort to define a `body-in-use-error` type and return that instead. And would make this much more self-documenting.

---

I haven't seen us use that idiom before in WASI of having singleton error variants; there's usually just a `result`.

---

oh, you're absolutely right: I had confused myself into thinking that if we move away from `error-code`, we need a specific replacement for it. But since the `result` here is only going to be `err` if the body is already taken out, I agree that that's not needed. Sorry for the noise.